File: src/leetcode/2026/04/Solution748.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

class Solution748 {
public:
    string shortestCompletingWord(string licensePlate, vector<string>& words) {
        int count[26] = {0};
        for (char c : licensePlate) {
            if (isalpha(c)) {
                count[tolower(c) - 'a']++;
            }
        }
        string result;
        for (const string& word : words) {
            int wordCount[26] = {0};
            for (char c : word) {
                wordCount[tolower(c) - 'a']++;
            }
            bool isCompleting = true;
            for (int i = 0; i < 26; ++i) {
                if (count[i] > wordCount[i]) {
                    isCompleting = false;
                    break;
                }
            }
            if (isCompleting) {
                if (result.empty() || word.length() < result.length()) {
                    result = word;
                }
            }
        }
        return result;
    }
};

int main() {
    Solution748 solution;

    string licensePlate1 = "1s3 PSt";
    vector<string> words1 = {"step", "steps", "stripe", "stepple"};
    cout << solution.shortestCompletingWord(licensePlate1, words1) << endl; // Output: "steps"

    string licensePlate2 = "1s3 456";
    vector<string> words2 = {"looks", "pest", "stew", "show"};
    cout << solution.shortestCompletingWord(licensePlate2, words2) << endl; // Output: "pest"

    return 0;
}

File: src/leetcode/2026/04/Solution748.java
public class Solution748 {
    public String shortestCompletingWord(String licensePlate, String[] words) {
        int[] count = new int[26];
        for (char c : licensePlate.toCharArray()) {
            if (Character.isLetter(c)) {
                count[Character.toLowerCase(c) - 'a']++;
            }
        }
        String result = null;
        for (String word : words) {
            if (isCompletingWord(word, count)) {
                if (result == null || word.length() < result.length()) {
                    result = word;
                }
            }
        }
        return result;
    }

    private boolean isCompletingWord(String word, int[] count) {
        int[] wordCount = new int[26];
        for (char c : word.toCharArray()) {
            if (Character.isLetter(c)) {
                wordCount[Character.toLowerCase(c) - 'a']++;
            }
        }
        for (int i = 0; i < 26; i++) {
            if (wordCount[i] < count[i]) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {
        Solution748 solution = new Solution748();

        String licensePlate = "1s3 PSt";
        String[] words = { "step", "steps", "stripe", "stepple" };
        String output = solution.shortestCompletingWord(licensePlate, words);
        System.out.println("Output: " + output);

        licensePlate = "1s3 456";
        words = new String[] { "looks", "pest", "stew", "show" };
        output = solution.shortestCompletingWord(licensePlate, words);
        System.out.println("Output: " + output);
    }
}
